### PR TITLE
<amp-consent> Support consentRequired and deprecate promptIfUnknownForGeo

### PR DESCRIPTION
**Pull request by micajuine-ho**
Task 3 & 4 of #25623.

- Support consentRequired with feature flag for legacy version and new version
- Deprecate `promptIfUnknownForGeo`. Migrate to `geoOverride` and `consentRequired`.
- Migration tests 
- Fixing old tests to support new migration behavior

Previous task(s):
1: #25676 (Merged)
2: #25689 (Merged)

### CODE REVIEW

File: extensions/amp-consent/0.1/consent-config.js [hunk 1]
```diff
@@ -150,6 +150,31 @@ export class ConsentConfig {
...
+    if (config['checkConsentHref'] && !config['consentRequired']) {
+      config['consentRequired'] = 'remote';
+    } else if (!config['consentRequired']) {
+      // Should we be conservative and make it true?
+      config['consentRequired'] = false;
+    }
+
+    const group = config['promptIfUnknownForGeoGroup'];
+    if (group) {
+      config['geoOverride'] = {
```
Comment (lannka) [inline extensions/amp-consent/0.1/consent-config.js]: also need to add
```
config['consentRequired'] = false;
```

File: extensions/amp-consent/0.1/consent-config.js [hunk 2]
```diff
@@ -150,6 +150,31 @@ export class ConsentConfig {
...
+        group: {
```
Comment (lannka) [inline extensions/amp-consent/0.1/consent-config.js]: this will not work. try `[group]`

File: extensions/amp-consent/0.1/consent-config.js [hunk 3]
```diff
@@ -150,6 +150,31 @@ export class ConsentConfig {
...
+
+    // consentRequired must not be null
+    if (config['checkConsentHref'] && !config['consentRequired']) {
+      config['consentRequired'] = 'remote';
+    } else if (!config['consentRequired']) {
+      // Should we be conservative and make it true?
+      config['consentRequired'] = false;
+    }
+
+    const group = config['promptIfUnknownForGeoGroup'];
```
Comment (lannka) [inline extensions/amp-consent/0.1/consent-config.js]: move this `promptIfUnknownForGeoGroup` handling before `consentRequired`

File: extensions/amp-consent/0.1/consent-config.js [hunk 4]
```diff
@@ -150,6 +150,31 @@ export class ConsentConfig {
       }
     }
 
+    userAssert(
```
Comment (lannka) [inline extensions/amp-consent/0.1/consent-config.js]: this assert might be not that useful. we can either have fallbacks, or fail individually.

**Comment by lannka** [comment]
you can add unit test for the migration now.

File: extensions/amp-consent/0.1/consent-config.js (hunk 3 above)
Comment (micajuine-ho) [reply extensions/amp-consent/0.1/consent-config.js]: What case are you concerned about?

File: extensions/amp-consent/0.1/consent-config.js (hunk 4 above)
Comment (micajuine-ho) [reply extensions/amp-consent/0.1/consent-config.js]: removed

File: extensions/amp-consent/0.1/consent-config.js (hunk 3 above)
Comment (micajuine-ho) [reply extensions/amp-consent/0.1/consent-config.js]: Actually, I see it now.

File: extensions/amp-consent/0.1/consent-config.js (hunk 1 above)
Comment (micajuine-ho) [reply extensions/amp-consent/0.1/consent-config.js]: Done

File: extensions/amp-consent/0.1/amp-consent.js [hunk 5]
```diff
@@ -435,6 +440,35 @@ export class AmpConsent extends AMP.BaseElement {
    * @return {!Promise<boolean>}
    */
   getConsentRequiredPromise_() {
+    if (isExperimentOn(this.win, 'amp-consent-geo-override')) {
```
Comment (lannka) [inline extensions/amp-consent/0.1/amp-consent.js]: if {!isExperimentOn...

Comment (micajuine-ho) [reply extensions/amp-consent/0.1/amp-consent.js]: oops... brain fart

File: extensions/amp-consent/0.1/consent-config.js [hunk 6]
```diff
@@ -151,6 +151,18 @@ export class ConsentConfig {
         }
       }
     }
+
+    // Migrate to geoOverride
+    const group = config['promptIfUnknownForGeoGroup'];
```
Comment (lannka) [inline extensions/amp-consent/0.1/consent-config.js]: add a TODO to `delete config['promptIfUnknownForGeoGroup']`

File: extensions/amp-consent/0.1/consent-config.js [hunk 7]
```diff
@@ -151,6 +151,18 @@ export class ConsentConfig {
...
+    if (group) {
```
Comment (lannka) [inline extensions/amp-consent/0.1/consent-config.js]: for safety, check `typeof group === "string"`

File: extensions/amp-consent/0.1/consent-config.js [hunk 8]
```diff
@@ -151,6 +151,18 @@ export class ConsentConfig {
...
+      config['consentRequired'] = false;
+      config['geoOverride'] = {
+        [group]: {
+          'consentRequired': true,
+        },
+      };
+    }
```
Comment (lannka) [inline extensions/amp-consent/0.1/consent-config.js]:
```js
} else if (config['consentRequired'] === undefined 
  && config['checkConsentHref']){
  config['consentRequired'] = 'remote';
}
```

File: extensions/amp-consent/0.1/consent-config.js [hunk 9]
```diff
@@ -151,6 +151,18 @@ export class ConsentConfig {
...
+
```
Comment (lannka) [inline extensions/amp-consent/0.1/consent-config.js]: then you can assert `config['consentRequired']` should not be `undefined` in `validateMergedGeoOverride_`

File: extensions/amp-consent/0.1/amp-consent.js [hunk 10]
```diff
@@ -439,11 +439,39 @@ export class AmpConsent extends AMP.BaseElement {
...
-      this.consentConfig_['checkConsentHref'] ||
-        this.consentConfig_['promptIfUnknownForGeoGroup'],
-      'neither checkConsentHref nor promptIfUnknownForGeoGroup is defined'
-    );
+    if (!isExperimentOn(this.win, 'amp-consent-geo-override')) {
+      return this.getConsentRequiredPromiseLegacy_();
+    }
+    let consentRequiredPromise = null;
+    if (
+      this.consentConfig_['consentRequired'] === true ||
```
Comment (lannka) [inline extensions/amp-consent/0.1/amp-consent.js]: typeof this.consentConfig_['consentRequired'] === 'boolean'

File: extensions/amp-consent/0.1/amp-consent.js [hunk 11]
```diff
@@ -439,11 +439,39 @@ export class AmpConsent extends AMP.BaseElement {
...
+      this.consentConfig_['consentRequired'] === false
+    ) {
+      consentRequiredPromise = Promise.resolve(
```
Comment (lannka) [inline extensions/amp-consent/0.1/amp-consent.js]: return Promise.resolve(...)

File: extensions/amp-consent/0.1/amp-consent.js [hunk 12]
```diff
@@ -439,11 +439,39 @@ export class AmpConsent extends AMP.BaseElement {
...
+        const remoteResponse = consentInfo['consentRequired'];
+
+        userAssert(
+          remoteResponse === true || remoteResponse === false,
+          'consentRequired from endpoint must be true or false'
+        );
+        return remoteResponse;
+      });
+    }
+    return consentRequiredPromise.then(required => {
```
Comment (lannka) [inline extensions/amp-consent/0.1/amp-consent.js]: the extra then block is not necessary

File: extensions/amp-consent/0.1/amp-consent.js [hunk 13]
```diff
@@ -439,11 +439,39 @@ export class AmpConsent extends AMP.BaseElement {
...
+    ) {
+      consentRequiredPromise = Promise.resolve(
+        this.consentConfig_['consentRequired']
+      );
+    } else {
+      consentRequiredPromise = this.getConsentRemote_().then(consentInfo => {
+        const remoteResponse = consentInfo['consentRequired'];
+
+        userAssert(
+          remoteResponse === true || remoteResponse === false,
```
Comment (lannka) [inline extensions/amp-consent/0.1/amp-consent.js]: typeof

File: extensions/amp-consent/0.1/amp-consent.js [hunk 14]
```diff
@@ -439,11 +439,39 @@ export class AmpConsent extends AMP.BaseElement {
...
+      this.consentConfig_['consentRequired'] === false
+    ) {
+      consentRequiredPromise = Promise.resolve(
+        this.consentConfig_['consentRequired']
+      );
+    } else {
+      consentRequiredPromise = this.getConsentRemote_().then(consentInfo => {
+        const remoteResponse = consentInfo['consentRequired'];
+
+        userAssert(
```
Comment (lannka) [inline extensions/amp-consent/0.1/amp-consent.js]: let's not throw. we can default to `true`

File: extensions/amp-consent/0.1/consent-config.js [hunk 15]
```diff
@@ -192,6 +210,12 @@ export class ConsentConfig {
    * @return {!JsonObject}
    */
   validateMergedGeoOverride_(mergedConfig) {
+    // TODO(micajuineho): remove `promptForGeoGroupUnknown`, after migration
+    userAssert(
+      mergedConfig['consentRequired'] !== undefined,
+      '`consentRequired` or `promptForGeoGroupUnknown` is required',
```
Comment (lannka) [inline extensions/amp-consent/0.1/consent-config.js]: no need to mention `promptForGeoGroupUnknown` as it's deprecated

File: extensions/amp-consent/0.1/test/test-consent-config.js [hunk 16]
```diff
@@ -107,6 +116,7 @@ describes.realWin('ConsentConfig', {amp: 1}, env => {
           'clientConfig': {
             'test': 'ABC',
           },
+          'consentRequired': false,
```
Comment (lannka) [inline extensions/amp-consent/0.1/test/test-consent-config.js]: do we have another test case that have this overridden to `true`?

Comment (micajuine-ho) [reply extensions/amp-consent/0.1/test/test-consent-config.js]: No, will add test.

**Pull request merged**